### PR TITLE
ASIM Audit schema parser with its sample and test data for Cisco Meraki.

### PR DESCRIPTION
**Pull request by jayeshprajapaticrest**
Required items, please complete
   
   Change(s):
   - Added ASIM Audit schema parser for Cisco Meraki based on Syslog data

   Reason for Change(s):
   - Initial version ASIM Audit Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+      | extend Parser = extract_all(@"(\d+.\d+)\s([\w\-\_]+)\s([\w\-\_]+)\s([\S\s]+)$", dynamic([1, 2, 3, 4]), LogMessage)[0]
+      | extend
+          Epoch = tostring(Parser[0]),
+          Substring = tostring(Parser[3]),
+          LogType = tostring(Parser[2])
+      | where (LogType == "events") and (Substring has_any ("vpn_connectivity_change", "status changed", "VRRP active", "VRRP passive") or Substring has_cs "Site-to-site" or Substring has_cs "Port")
+      | extend EpochTimestamp = split(Epoch, ".")
+      | extend EventStartTime = unixtime_seconds_todatetime(tolong(EpochTimestamp[0]))
+      | extend EventEndTime = EventStartTime 
+      | where (isnull(starttime) or EventStartTime >= starttime)
```
Comment (devikamehra) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:88]: Time filters should always be first

**Review by devikamehra — DISMISSED** [review]
For vim parser, filter->parse->map is not followed. Please ensure we perform required pre and post filtering for best results and also ensure that they are early in the code.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+      "purging ISAKMP-SA", "Partial", "Delete",
+      "purged ISAKMP-SA", "Success", "Delete",
+      "ISAKMP-SA deleted", "Success", "Delete",
+      "IPsec-SA request queued due to no phase 1 found", "Failure", "Execute",
+      "failed to get sainfo", "Failure", "Execute",
+      "failed to pre-process ph2 packet", "Failure", "Execute",
+      "phase2 negotiation failed due to time up waiting for phase1", "Failure", "Execute",
+      "initiate new phase 1 negotiation", "Success", "Execute",
+      "ISAKMP-SA established", "Success", "Execute",
+      "initiate new phase 2 negotiation", "Partial", "Execute",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+ParserName: ASimAuditEventCiscoMeraki
+EquivalentBuiltInParser: _ASim_AuditEvent_CiscoMeraki
+ParserParams:
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
+  let lookUpTable = datatable(Operation: string, EventResult: string, EventType: string)
+    [
+      "vpn_connectivity_change", "Success", "Set",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      "purging ISAKMP-SA", "Partial", "Delete",
+      "purged ISAKMP-SA", "Success", "Delete",
+      "ISAKMP-SA deleted", "Success", "Delete",
+      "IPsec-SA request queued due to no phase 1 found", "Failure", "Execute",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+      "failed to get sainfo", "Failure", "Execute",
+      "failed to pre-process ph2 packet", "Failure", "Execute",
+      "phase2 negotiation failed due to time up waiting for phase1", "Failure", "Execute",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+      | where not(disabled)
+          and LogMessage has "events"
+          and (LogMessage has_any ("vpn_connectivity_change", "status changed", "VRRP active", "VRRP passive") or LogMessage has_cs "Site-to-site" or LogMessage has_cs "Port")
+      | extend Parser = extract_all(@"(\d+.\d+)\s([\w\-\_]+)\s([\w\-\_]+)\s([\S\s]+)$", dynamic([1, 2, 3, 4]), LogMessage)[0]
+      | extend
+          Substring = tostring(Parser[3]),
+          LogType = tostring(Parser[2])
+      | where (LogType == "events") and (Substring has_any ("vpn_connectivity_change", "status changed", "VRRP active", "VRRP passive") or Substring has_cs "Site-to-site" or Substring has_cs "Port")
+      | parse-kv Substring as (type: string, peer_contact: string, connectivity: string) with (pair_delimiter=" ", kv_delimiter="=")
+      | parse Substring with * " deleted " temp_deletedSrcIp "-" temp_deletedTargetIp " " temp_restmessage
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                          Operation == "ISAKMP-SA established",
+                          temp_estTargetIp,
+                          Operation == "initiate new phase 2 negotiation",
+                          temp_negotiationTargetIp,
+                          Operation == "IPsec-SA established",
+                          temp_tunnelTargetIp,
+                          ""
+                      )
+      | extend
+          temp_srcipport = iff(temp_srcipport has "[" and temp_srcipport !has ":", replace_strings(temp_srcipport, dynamic(['[', ']']), dynamic([':', ''])), temp_srcipport),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 1 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:87]: Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+Description: |
+  This ASIM parser supports normalizing Cisco Meraki to the ASIM Audit Event schema.
+ParserName: ASimAuditEventCiscoMeraki
+EquivalentBuiltInParser: _ASim_AuditEvent_CiscoMeraki
+ParserParams:
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
+  let lookUpTable = datatable(Operation: string, EventResult: string, EventType: string)
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Done.
EventResultDetails field we have added based on condition as it's not possible to add in lookup.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+          | project-rename LogMessage =  Message
+          ),
+          (
+          Syslog
+          | project-rename LogMessage =  SyslogMessage
+          )
+      | where not(disabled)
+          and LogMessage has "events"
+          and (LogMessage has_any ("vpn_connectivity_change", "status changed", "VRRP active", "VRRP passive") or LogMessage has_cs "Site-to-site" or LogMessage has_cs "Port")
+      | extend Parser = extract_all(@"(\d+.\d+)\s([\w\-\_]+)\s([\w\-\_]+)\s([\S\s]+)$", dynamic([1, 2, 3, 4]), LogMessage)[0]
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml:64]: Event format1:
2019-07-16 04:00:30 [Local0.Info](http://local0.info/) 172.40.20.177 1 1561036264.565291108 australia_sydney security_event security_filtering_file_scanned url=http://tlu123.dl123.delivery.mp.microsoft.com/filestreamingservice/files/36ec4eb7-46dd-4aeb-990e-b6d32f7ed567?P1=1561036860& P2=402&P3=2&P4=Zj3qRDR5CbzfWlP8BuYg%2bUlTon0XE774ExEEquiawstLAJ2%2bQm3OoWLcwz3HBt8qp3r3buVRVoT5BQcUCcNlXw%3d%3d src=[172.16.10.98:64160](http://172.16.10.98:64160/) dst=[200.188.210.42:180](http://200.188.210.42:180/) mac=20:1C:BC:B2:0F:20 name='' sha256093e4fc218b27e58e2fede7b8cb044d48d66995ae785bbc186a9df5ae08ca4f7 disposition=malicious action=block

Event format2:
<134>1 1685956046.281934865 CDSAPF3_11 events type=association radio='1' vap='3' client_mac='F2:ED:56:6A:98:6A' last_known_client_ip='172.20.12.194' band='5' channel='104' rssi='38' identity='zuber.chataiwala' aid='1305860724'


We double checked and we have the different event formats as above so we are not able to do the parsing as above to get the required fields.

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+                          Substring has "IPsec-SA established",
+                          "IPsec-SA established",
+                          Substring has "status changed",
+                          "port status change",
+                          Substring has_cs "Port" and Substring has "STP BPDU",
+                          "spanning-tree guard state change",
+                          Substring has_cs "Port" and Substring has "STP role",
+                          "spanning-tree interface role change",
+                          Substring has_any("VRRP active", "VRRP passive"),
+                          "VRRP transition",
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: value "VRRP transition" is not coming in data but we are assigning value to it according to the type mentioned in Cisco Meraki [document](https://documentation.meraki.com/General_Administration/Monitoring_and_Reporting/Syslog_Event_Types_and_Log_Samples).
image

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Can you validate if this suggestion has been applied? I couldn't find

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: can you please check these EventResultDetailsValues output from the parser?
image
image

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please act on @vakohl's comments. Thanks

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml]: Yes. Done

File: Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/ASimAuditEventCiscoMeraki.yaml:65]: Replied websession PR.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+      | extend Parser = extract_all(@"(\d+.\d+)\s([\w\-\_]+)\s([\w\-\_]+)\s([\S\s]+)$", dynamic([1, 2, 3, 4]), LogMessage)[0]
+      | extend
+      Epoch = tostring(Parser[0])
+      | extend EpochTimestamp = split(Epoch, ".")
+      | extend EventStartTime = unixtime_seconds_todatetime(tolong(EpochTimestamp[0]))
+      | extend LogType = tostring(Parser[2]),
+          Substring = tostring(Parser[3])
+      | where LogType == "events"
+      and (array_length(eventtype_in) == 0 or LogMessage has_any (eventtype_in))
+      and (eventresult == "*" or LogMessage has eventresult);
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:94]: This filter won't work as LogMessage won't contain these values. Same filter being used at line 167, can we remove this one?
Please test all filter once.

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+Normalization:
+  Schema: AuditEvent
+  Version: '0.1'
+References:
+- Title: ASIM Audit Event Schema
+  Link: https://aka.ms/ASimAuditEventDoc
+- Title: ASIM
+  Link: https://aka.ms/AboutASIM
+- Title: Cisco Meraki Documentation
+  Link: https://documentation.meraki.com/
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:16]: can we add direct link of the page which contain Audit logs related info?

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+          and (LogMessage has_any ("vpn_connectivity_change", "status changed", "VRRP active", "VRRP passive") or LogMessage has_cs "Site-to-site" or LogMessage has_cs "Port")
+      | extend Parser = extract_all(@"(\d+.\d+)\s([\w\-\_]+)\s([\w\-\_]+)\s([\S\s]+)$", dynamic([1, 2, 3, 4]), LogMessage)[0]
+      | extend
+      Epoch = tostring(Parser[0])
+      | extend EpochTimestamp = split(Epoch, ".")
+      | extend EventStartTime = unixtime_seconds_todatetime(tolong(EpochTimestamp[0]))
+      | extend LogType = tostring(Parser[2]),
+          Substring = tostring(Parser[3])
+      | where LogType == "events"
+      and (array_length(eventtype_in) == 0 or LogMessage has_any (eventtype_in))
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:93]: LogMessage won't match with EventType value, same filter we are using at 172, can we remove this one?

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+  Link: https://aka.ms/ASimAuditEventDoc
+- Title: ASIM
+  Link: https://aka.ms/AboutASIM
+- Title: Cisco Meraki Documentation
+  Link: https://documentation.meraki.com/
+Description: |
+  This ASIM parser supports normalizing Cisco Meraki logs to the ASIM Audit Event normalized schema. Cisco Meraki events are generated from network activity and security events from Meraki devices such as firewalls, switches, and access points. These logs are captured through the Cisco Meraki Sentinel connector which uses a Linux agent to collect logs in Syslog format.
+ParserName: vimAuditEventCiscoMeraki
+EquivalentBuiltInParser: _Im_AuditEvent_CiscoMeraki
+ParserParams:
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:21]: we are missing few audit filters. please include them too. If we don't have values in those fields, we can check for fitlers containing null or '*'
https://learn.microsoft.com/en-us/azure/sentinel/normalization-schema-audit#filtering-parser-parameters

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+          temp_srcipport = trim("'", temp_srcipport),
+          temp_targetipport = trim("'", temp_targetipport)
+      | extend 
+          temp_srcipport = trim('"', temp_srcipport),
+          temp_targetipport = trim('"', temp_targetipport)
+      | parse temp_srcipport with * "[" temp_srcip "]:" temp_srcport
+      | extend SrcIpAddr = iff(temp_srcipport has ".", split(temp_srcipport, ":")[0], coalesce(temp_srcip, temp_srcipport))
+      | extend
+      temp_SrcMatch=has_any_ipv4_prefix(SrcIpAddr, srcipaddr_has_any_prefix)
+      | where array_length(srcipaddr_has_any_prefix) == 0 or isnotempty(temp_SrcMatch)
```
Comment (vakohl) [inline Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:191]: src ip filter doesn't work currently. try:
| where array_length(srcipaddr_has_any_prefix) == 0 or temp_SrcMatch == true

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please act on @vakohl's comments. Also kql validations failed please see below screenshot for the same. Thanks
image

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 11 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 12 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 13 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 14 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml:21]: Done

File: Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml (hunk 15 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimAuditEvent/Parsers/vimAuditEventCiscoMeraki.yaml]: Done

**Comment by v-atulyadav** [comment]
Hi @jayeshprajapaticrest, please act on @vakohl's comments. Also kql validations failed please see below screenshot for the same. Thanks
image

**Comment by vakohl** [comment]
> Hi @jayeshprajapaticrest, please act on @vakohl's comments. Also kql validations failed please see below screenshot for the same. Thanks image

@v-atulyadav kql validation issues are fixed

**Review by vakohl — APPROVED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**